server/src/main/kotlin/org/javacs/kt/CompilerClassPath.kt

```diff
@@ -36,9 +38,18 @@ class CompilerClassPath(private val config: CompilerConfiguration) : Closeable {
         if (updateClassPath) {
             val newClassPath = resolver.classpathOrEmpty
             if (newClassPath != classPath) {
-                syncPaths(classPath, newClassPath, "class path")
+                synchronized(classPath) {
+                    syncClassPathEntries(classPath, newClassPath, "class path")
+                }
                 refreshCompiler = true
             }
+
+            async.compute {
+                val newClassPathWithSources = resolver.classpathWithSources
+                synchronized(classPath) {
+                    syncClassPathEntries(classPath, newClassPathWithSources, "class path")
+                }
+            }
```

Comment:
I think, we have to be careful here, not to introduce any race conditions. The `Compiler` instance expects a full class path AFAICT, so perhaps we should keep it as a synchronous operation?

Reply:
You're right. Although making it synchronous might significantly delay the language server startup, especially for bigger projects. This should only happen on the first time (on subsequent startups, no downloading will be required), but I don't know how I feel about this

Reply:
Ideally, we should make the update atomic. Not sure how that will affect other parts of the code though

Reply:
@fwcd Any thoughts on this?

Reply:
Not sure, how would the language server be able to provide be able to compile the project (and provide useful diagnostics) without the full class path?

Reply:
> The base class path is computed synchronously and just the source jars are fetched asynchronously, right?

That is correct. My motivation to make it async is that source JARs are not a "vital" feature IMO. And for large projects it might take some time to download all the JARs the first time.